fixes #1558: Wrap terms with evaluation only if we have a regex that fail normalization

ivakegg opened the pull request.

ivakegg committed:
fixes #1558: Wrap terms with evaluation only if we have a regex that fails normalization.

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTerms.java:
```diff
@@ -259,6 +261,14 @@ protected JexlNode expandNodeForNormalizers(JexlNode node, Object data) {
                             }
+                            // if this was a regex, then lets assume that the regex could not be decoded enough to determine how to match against the index.
+                            // in this case we need to force this term to be evaluation only (i.e. cannot be done against an index)
+                            if (node instanceof ASTNRNode || node instanceof ASTERNode) {
```
Can we expand the test case below to also hit the ASTNRNode?

ivakegg replied:
I did, and I also realized that we were not handling them correctly.  Note that the NR and NE cases are not actually ever exercised  because all NE and NR nodes are turned into EQ and ER nodes surrounded by a negation before we get here.

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTerms.java:
```diff
@@ -274,6 +284,13 @@ protected JexlNode expandNodeForNormalizers(JexlNode node, Object data) {
                         nodeToReturn = JexlNodeFactory.buildUntypedNewLiteralNode(node, fieldName, literal);
                     }
                     
+                    // if we have an unnormalizable regex in the mix, then wrap with evaluation only
+                    // this is ok even if we have a mix of normalizations that worked vs not because
+                    // if one of a union is evaluation only, then they all must be (see executability
+                    // visitor for more understanding)
```
Does this imply we need to push whole  unions into EvalOnly? `(A or B or (eval & C)) -> (eval & (A or B or C))`

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTerms.java:
```diff
@@ -274,6 +284,13 @@ protected JexlNode expandNodeForNormalizers(JexlNode node, Object data) {
                         nodeToReturn = JexlNodeFactory.buildUntypedNewLiteralNode(node, fieldName, literal);
                     }
                     
+                    // if we have an unnormalizable regex in the mix, then wrap with evaluation only
+                    // this is ok even if we have a mix of normalizations that worked vs not because
+                    // if one of a union is evaluation only, then they all must be (see executability
+                    // visitor for more understanding)
```
Does this imply we need to push whole  unions into EvalOnly? `(A or B or (eval & C)) -> (eval & (A or B or C))`

ivakegg replied:
So this is the case were we have one field and multiple normalizations there of in which one or more of them failed.  In this case I thought it reasonable to send the whole expression to be eval only.  For example:
```
FIELD =~ '.*AE3.*'
where FIELD is a String and a Number type
becomes:  ((_Eval_ = true ) && ( ( FIELD =~ '.*ae3.*' ) || ( FIELD =! '.*AE3.*' ) ) )
```
If I cannot run one of them against the index, then the entire union cannot be run against the index.

hlgp commented on warehouse/query-core/src/test/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTermsTest.java:
```diff
@@ -490,6 +488,27 @@ public void testDelayedPredicates() throws ParseException {
+        
+        config.setQueryFieldsDatatypes(dataTypes);
+        
+        // this tests for the successful normalization as a simply number can be normalized as a regex
```
simple number

ivakegg replied:
ok

hlgp commented on warehouse/query-core/src/test/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTermsTest.java:
```diff
@@ -490,6 +488,27 @@ public void testDelayedPredicates() throws ParseException {
     }
     
+    @Test
+    public void testFailedRegexNormalizers() throws ParseException {
```
Maybe a few more test cases? 
Consider the case when the field is index only.
Are we going to update the HitListArithmetic where it uses the normalized form out of the tuple? Or is that out of the scope of this issue?  Currently if there's an event with FRUIT:BANANA and WEIGHT:40, the query 
FRUIT == 'banana' AND WEIGHT =~ '4.\*'
will result in the correct plan, FRUIT == 'banana' && (( _Eval_=true) && WEIGHT =~ '4.\*')
but will try to compare the pattern, 4.* to +bE4, which doesn't work.  Maybe I'm missing something.

ivakegg replied:
looking at fixing the HitListArithmetic to be consistent with the DefaultArithmetic.

ivakegg committed:
re #1558: Review comments
* Fixed the hit list arithmetic to be able to apply regex against the
  base value and the normalized value.  Added test cases to ensure
  consistency with the DefaultArithmetic.

ivakegg committed:
re #1558: Ensure that NE and NR nodes are appropriatly handled

apmoriarty committed:
Merge branch 'integration' into feature/numericRegex

apmoriarty committed:
Merge branch 'integration' into feature/numericRegex

hgklohr committed:
Merge branch 'integration' into feature/numericRegex
